### PR TITLE
Feat/237 make individualrecipescreenviewmodel

### LINKED ISSUE #237: make IndividualRecipeScreenViewModel
We need to add a view model for the individualRecipeScreen

### PR DESCRIPTION
**Pull request by AleMerRom**
# PR
## Description
In this PR, I have added the viewModel for the individualRecipeScreen

## **Main Changes**
1. **IndividualRecipeViewModel.kt**:
This is the viewModel for the individualRecipeScreen. Inside it has functions to access all necessary information from the selected recipe.

2. ** IndividualRecipe folder**:
Because the backend factorization has been divided between the different team members, we need to keep the old files to make sure the app will work. 
- **OldIndividualRecipe.kt** => the old IndividualRecipe
- **IndividualRecipe.kt** => the new and perfected file:
   - used the functions form the viewModel
   - slightly changed the logic
   - added an easter egg screen. It is only accessible if no recipe is selected. In this screen,  I add  the imaged **how_did_we_get_here.png**
   - I have also added 2 preview functions to test the UI and check the viewModel works. One is for an unselected recipe and the other for normal use.
Screenshot_2024-12-04-20-50-58-714_com android shelfLife 1

3. **RecipeRepositoryFirestore.kt** & **RecipeRepository.kt** :
We need to keep track of a selectedRecipe. So I have added these functions to manage the selection of a recipe

### **Modified**:
- **EndToEndM2Test.kt**
- **IndividualRecipeTest.kt**
- **MainActivity.kt**
- **RecipeRepository.kt**
- **RecipeRepositoryFirestore.kt**
- **IndividualRecipe.kt**
- **OldIndividualRecipe.kt**
- **Recipes.kt**
### **Added**:
- **IndividualRecipeViewModel.kt**
- **how_did_we_get_here.png**

### CODE REVIEW

File: app/src/main/java/com/android/shelfLife/model/recipe/RecipeRepository.kt [hunk 1]
```diff
@@ -1,6 +1,12 @@
 package com.android.shelfLife.model.recipe
 
+import kotlinx.coroutines.flow.StateFlow
+
 interface RecipeRepository {
+
+  val recipes: StateFlow<List<Recipe>>
+  val selectedRecipe: StateFlow<Recipe?>
```
Comment (alextheprophecy) [inline app/src/main/java/com/android/shelfLife/model/recipe/RecipeRepository.kt:8]: ## To be further discussed
I believe it was decided with some members that the selectedRecipe state should be saved in the Recipe repository. 
I dont feel like it belongs here, since the selection is purely a visual (UI) state and has nothing to do with the repository/model.

It is fine for now, **no breaking change** but I think this should be adressed later

File: app/src/main/java/com/android/shelfLife/ui/recipes/IndividualRecipe/IndividualRecipe.kt [hunk 2]
```diff
@@ -0,0 +1,259 @@
...
+  recipeRepository.selectRecipe(recipe)
+  Log.i("AAAAAAAAA", "3")
+
+  val individualRecipeViewModel = viewModel { IndividualRecipeViewModel(recipeRepository) }
+  Log.i("AAAAAAAAA", "4")
+
+  // Render the IndividualRecipeScreen with a null selectedRecipe
+  IndividualRecipeScreen(
+      navigationActions = navigationActions, individualRecipeViewModel = individualRecipeViewModel)
+}
```
Comment (alextheprophecy) [inline app/src/main/java/com/android/shelfLife/ui/recipes/IndividualRecipe/IndividualRecipe.kt:259]: ## Nice use of Preview to help debugging
This was very useful as a PR reviewer to see the screens, and debug quickly
## Should this be pushed to main however? 
This code is wont actually be used in the app- its just a **Test**
I feel like this should be covered by UI tests (i.e checking the easter egg image and so for the edge cases).

File: app/src/main/java/com/android/shelfLife/ui/recipes/IndividualRecipe/IndividualRecipe.kt [hunk 3]
```diff
@@ -0,0 +1,259 @@
...
+                modifier =
+                    Modifier.padding(8.dp)
+                        .fillMaxSize()
+                        .verticalScroll(rememberScrollState()) // Enable vertical scrolling
+                ) {
+                  // Display the recipe image (placeholder for now)
+                  Image(
+                      painter = painterResource(R.drawable.google_logo),
+                      contentDescription = "Recipe Image",
+                      modifier = Modifier.width(537.dp).height(159.dp).testTag("recipeImage"),
```
Comment (alextheprophecy) [inline app/src/main/java/com/android/shelfLife/ui/recipes/IndividualRecipe/IndividualRecipe.kt:104]: ## Not friendly for small screens
 Would it be possible to use percentages here instead? So it properly scales with device width
```suggestion
                      modifier = Modifier.fillMaxWidth(0.6f)(537.dp).height(159.dp).testTag("recipeImage"),
```

File: app/src/main/java/com/android/shelfLife/viewmodel/recipes/IndividualRecipeViewModel.kt [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+    }
+  }
+
+  fun getRecipeName(): String {
+    /**
+     * Returns the name of the selected recipe.
+     *
+     * @return The name of the recipe as a string.
+     */
+    return selectedRecipe!!.name
```
Comment (alextheprophecy) [inline app/src/main/java/com/android/shelfLife/viewmodel/recipes/IndividualRecipeViewModel.kt:35]: ## Error potential
Since selectedRecipe can be null by definition `Recipe?`, you should not force it with !!.
Instead, check if null
```suggestion
    return if selectedRecipe==null "" else selectedRecipe!!.name
```

Comment (alextheprophecy) [reply app/src/main/java/com/android/shelfLife/viewmodel/recipes/IndividualRecipeViewModel.kt:35]: This applies for all the rest of the functions

File: app/src/main/java/com/android/shelfLife/viewmodel/recipes/IndividualRecipeViewModel.kt [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+      selectedRecipeIsNonEmpty = false
+    }
+  }
+
+  fun getRecipeName(): String {
+    /**
+     * Returns the name of the selected recipe.
+     *
+     * @return The name of the recipe as a string.
+     */
```
Comment (alextheprophecy) [inline app/src/main/java/com/android/shelfLife/viewmodel/recipes/IndividualRecipeViewModel.kt:34]: nit-picky but it hurts my eyes- this comment should go just before the function defintion

**Review by alextheprophecy — APPROVED** [review]
## Overall nice PR
Good job on the little easter egg, the app is missing a bit of fun and gamification so its a great addition in the right direction.

## Potential changes
Some style and code restructuring necessary as mentioned. Also the functions in `IndividualRecipeViewModel` could be error prone
